### PR TITLE
Add support for parsing neuron core resource limit and pass it as ray…

### PR DESCRIPTION
**Pull request by mounchin**
## Why are these changes needed?

The PR is needed to parse the neuroncore resource limits fro the pod spec and pass them to `ray start` command as it expects.
Expected format: https://docs.ray.io/en/latest/ray-core/scheduling/accelerators.html

## Related issue number

https://github.com/ray-project/ray/issues/44361

## Checks

- [x] I've made sure the tests are passing.
- Testing Strategy
   - [x] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [inline ray-operator/controllers/ray/common/pod.go:762]: Can you put `aws.amazon.com/neuroncore` into a constant?

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [inline ray-operator/controllers/ray/common/pod.go:763]: This would be the first hardware accelerator (outside GPUs) that we auto detect in container resource and pass into the `ray start` command as custom resource. I can see the appeal of doing this but it could also mean we end up supporting an ever growing list of custom resources in kuberay.

My biggest question here is whether having to specify the custom resource in `startParams` is a big enough pain point for KubeRay to auto detect custom resources.

Comment (andrewsykim) [reply ray-operator/controllers/ray/common/pod.go:763]: Thinking about it more, it's probably worth doing for the well-known accelerators at least. Can we generalize the implementation so it's easy to add new accelerators here? cc @ryanaoleary for TPUs

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:763]: Native support for neuron and other accelerators already exists for Ray's VM cluster-launchers.
See discussion in https://github.com/ray-project/ray/issues/44361

This PR would move KubeRay closer to parity with the VM-based solution.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:763]: I agree that there needs to be some styling here that could generalize to other hardware,
again along the same lines as what exists for Ray's VM support.

File: ray-operator/controllers/ray/common/pod.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go]: Made it a constant

File: ray-operator/controllers/ray/common/pod.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go]: I have generalized code for adding other accelerator types using a map, let me know if you see any other issues.

**Review by DmitriGekhtman — APPROVED** [review]
Looks good to me.
I recommend getting approval from one of the more active KubeRay maintainers (e.g. @andrewsykim or @kevin85421 ) prior to merge.

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [inline ray-operator/controllers/ray/common/pod.go:36]: Can we also track the Ray resource name as a constant?
```
NeuronCoreContainerResourceName = "aws.amazon.com/neuroncore"
NeuronCoreRayResourceName = "neuron_cores"
```

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [inline ray-operator/controllers/ray/common/pod.go:765]: This comment and `break` below is now no longer true. I'm not sure how common it is to have nodes that have both GPUs and custom accelerators, but I think the `break` can be removed

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [inline ray-operator/controllers/ray/common/pod.go:759]: Does it make sense to gate the custom accelerator check if `num-gpus` doesn't exist? Maybe we need to put the custom accelerator check outside this if block

File: ray-operator/controllers/ray/common/pod.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go:765]: The comment can be removed, but I think `break` is needed to match the first `resourceKey` and then exit of the loop, instead of parsing both and passing them to the `ray start`. This is assuming that either `num-gpus` can exist or the custom accelerator types.

https://docs.ray.io/en/latest/ray-core/scheduling/accelerators.html

File: ray-operator/controllers/ray/common/pod.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go:759]: Again, this is based on the assumption that either `num-gpus` or the custom accelerator type can exit on the node.  So that only check for other resource types if `num-gpus` does not exist.

File: ray-operator/controllers/ray/common/pod.go
Comment (andrewsykim) [reply ray-operator/controllers/ray/common/pod.go:765]: Is there somewhere in the docs that says you can't specify `--num-gpus` and a custom accelerator?

I think my main point here is that if a user specifies `--resources` in rayStartParams today with AWS neuron cores, it should override the new behavior of parsing the container resource limits. Right now the behavior only applies if `--num-gpus` is not specified which I think is incorrect.

Comment (andrewsykim) [reply ray-operator/controllers/ray/common/pod.go:765]: Consider changing the code to something like this:
```go
if _, ok := rayStartParams["num-gpus"]; !ok {
    // Scan for resource keys ending with "gpu" like "nvidia.com/gpu".
    for resourceKey, resource := range resource.Limits {
	    resourceKeyString := string(resourceKey)
	    if strings.HasSuffix(resourceKeyString, "gpu") && !resource.IsZero() {
		    rayStartParams["num-gpus"] = strconv.FormatInt(resource.Value(), 10)
		    // For now, only support one GPU type. Break on first match.
		    break
	    }
    }
}

if _, ok := rayStartParams["resources"]; !ok {
  for resourceKey, resource := range resource.Limits {
      resourceKeyString := string(resourceKey)
      if rayResourceName, ok := customAcceleratorToRayResourceMap[resourceKeyString]; ok && !resource.IsZero() {
            if err := addCustomAcceleratorToResourcesIfNotExists(rayStartParams, rayResourceName, resource.Value()); err != nil {
                log.Error(err, fmt.Sprintf("failed to add %s to resources", rayResourceName))
            }
            // For now, only support one custom accelerator type. Break on first match.
            break
      }
  }
}
```

Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go:765]: The documentation does not say that it supports both `gpu` and a custom accelerator in combination either. So I am not sure if we should allow parsing both. Wdyt?

And for the code change you suggested above, we do not want to override resources if its specified already? Is that right? Can other params be passed as part of resources or just custom accelerator types, and we do not want to add eg: `neuron_cores` based on the resourceLimits specified?

Comment (andrewsykim) [reply ray-operator/controllers/ray/common/pod.go:765]: > And for the code change you suggested above, we do not want to override resources if its specified already? Is that right? Can other params be passed as part of resources or just custom accelerator types, and we do not want to add eg: neuron_cores based on the resourceLimits specified?

I suggest following the same pattern as `num-gpus` for custom resources. We should only set it if not specified in rayStartParam but specified in as a container resource limit. If specified in start params, we shouldn't override it.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:765]: > The documentation does not say that it supports both gpu and a custom accelerator in combination either.

Technically, Ray wouldn't have an issue with that.
But it would be unusual for a cloud provider to support attaching a provider-specific chip and an nvidia gpu to a single VM.

File: ray-operator/controllers/ray/common/pod.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod.go]: Moved it into another `if` block

File: ray-operator/controllers/ray/common/pod_test.go
Comment (mounchin) [reply ray-operator/controllers/ray/common/pod_test.go:1200]: Definitely!! From [this](https://github.com/ray-project/kuberay/pull/2409#discussion_r1783606237) conversation I thought we do not want to add/update `resources`, if its already specified. Can you answer below questions?
- If `neuron_cores` exist in the resources, do we want to override with the value in container limits?
- If `resources` exist and `neuron_cores` does not exist, do we want to parse the container limits and add them to the resources?
- If `resources` does not exist, do we want to add the `neuron_cores`?

I'm assuming the answers are No, Yes and Yes.

**Review by andrewsykim — APPROVED** [review]

File: ray-operator/controllers/ray/common/pod_test.go
Comment (andrewsykim) [reply ray-operator/controllers/ray/common/pod_test.go:1200]: Yeah that looks right to me. Can you open a follow up PR to support that?

Comment (mounchin) [reply ray-operator/controllers/ray/common/pod_test.go:1200]: Yes I will open a followup PR. Thanks for merging the change.

**Comment by ryanaoleary** [comment]
@mounchin Can you explain the use-case for this change a bit more? I was under the impression that the Raylet was already aware of custom resources/accelerators passed to the Pod container resources. For example, even though the generated Ray start command for a RayCluster with a TPU worker group might look like this:
```
# ray head
ray start --head  --dashboard-host=0.0.0.0  --metrics-export-port=8080  --block  --dashboard-agent-listen-port=52365  --num-cpus=8  --memory=40000000000
# ray worker with TPU
ray start  --address=vllm-tpu-head-svc.default.svc.cluster.local:6379  --metrics-export-port=8080  --block  --dashboard-agent-listen-port=52365  --num-cpus=100  --memory=100000000000 
```
The RayCluster controller still parses the [resources ](https://github.com/ray-project/kuberay/blob/bf21d2d01cf1c931136d869d2c8168aed07bc68c/ray-operator/controllers/ray/raycluster_controller.go#L1243) and shows TPUs available when running `ray list nodes`:
```
Total: 3

Table:
------------------------------
    NODE_ID                                                   NODE_IP      IS_HEAD_NODE    STATE    STATE_MESSAGE    NODE_NAME    RESOURCES_TOTAL                      LABELS
 0  2fc809f11e242261a66ab89ab33e5849639ef6be69543e6d038f80b4  10.76.0.196  False           ALIVE                     10.76.0.196  CPU: 100.0                           ray.io/node_id: 2fc809f11e242261a66ab89ab33e5849639ef6be69543e6d038f80b4
                                                                                                                                  TPU: 4.0
                                                                                                                                  TPU-v5litepod-8-head: 1.0
                                                                                                                                  accelerator_type:TPU-V5LITEPOD: 1.0
                                                                                                                                  memory: 93.132 GiB
                                                                                                                                  node:10.76.0.196: 1.0
                                                                                                                                  object_store_memory: 27.906 GiB
                                                                                                                                  tpu-group-0: 1.0
 1  6534925164aa45e434f75115289bddb2710671c00bfc0fa6cbf74652  10.76.0.229  False           ALIVE                     10.76.0.229  CPU: 100.0                           ray.io/node_id: 6534925164aa45e434f75115289bddb2710671c00bfc0fa6cbf74652
                                                                                                                                  TPU: 4.0
                                                                                                                                  accelerator_type:TPU-V5LITEPOD: 1.0
                                                                                                                                  memory: 93.132 GiB
                                                                                                                                  node:10.76.0.229: 1.0
                                                                                                                                  object_store_memory: 27.906 GiB
                                                                                                                                  tpu-group-0: 1.0
 2  e9cdbf0dc323d1a1a434cc319469248d1ad0e63060169bbe31301cfd  10.76.0.141  True            ALIVE                     10.76.0.141  CPU: 8.0                             ray.io/node_id: e9cdbf0dc323d1a1a434cc319469248d1ad0e63060169bbe31301cfd
                                                                                                                                  memory: 37.253 GiB
                                                                                                                                  node:10.76.0.141: 1.0
                                                                                                                                  node:__internal_head__: 1.0
                                                                                                                                  object_store_memory: 11.076 GiB
```
i.e. the RayCluster status updates based on the Pod container resources even if they're not passed to the ray start command. Is there something I'm missing here that this change enables with Ray?

**Comment by mounchin** [comment]
@ryanaoleary you can read more about the use-case [here](https://github.com/ray-project/ray/issues/44361#issuecomment-2274999901). Let me know if you still think this change to parse pod limits and pass them as resources is not needed.

**Comment by mounchin** [comment]
> @mounchin have you tested that this change resolves the autoscaling issue mentioend in [ray-project/ray#44361 (comment)](https://github.com/ray-project/ray/issues/44361#issuecomment-2274999901)?

We need changes on the EKS karpenter side to support scaling based on neuron cores to test E2E, awaiting a PR merge on karpenter side.

**Comment by andrewsykim** [comment]
When do you think the validation will be done? Ideally we want to know before the v1.3 release in case we need to revert detection of custom accelerators (if it's not resolving the autoscaling issue)

**Comment by mounchin** [comment]
Whats the ETA for the v1.3 release, I will check with the issue owner to see if this can be validated with out using karpenter and just use auto-scaler. Will let you know.

**Pull request merged**
